### ISSUE TITLE
[HOLD for payment 2024-03-19] HIGH: [$500] Magic link - Can't sign in using magic link after going to public room via deeplink

### ISSUE THREAD

**Issue by isagoico** [Expensify/App#36812]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:** 1.4.42.0

**Reproducible in staging?:** Yes 
**Reproducible in production?:** Yes

**Issue reported by:** Applause -Internal team
**Slack conversation:** Found while executing tests for https://expensify.slack.com/archives/C9YU7BX5M/p1708359665549809

## Action Performed:
1. Sign out (if needed) from the app (And kill the app)
2. Navigate to this public room from another app: https://staging.new.expensify.com/r/6776919265133947
3. Click on sign in
4. Open gmail app and click on the sign in magic link

## Expected Result:
App should automatically sign the user in

## Actual Result:
App doesn't automatically sign the user in after clicking on magic link. user has to use magic code to sign in

## Workaround:
N/A

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [x] Android: Native
- [ ] Android: mWeb Chrome
- [ ] iOS: Native
- [ ] iOS: mWeb Safari
- [ ] MacOS: Chrome / Safari
- [ ] MacOS: Desktop

## Screenshots/Videos

https://github.com/Expensify/App/assets/44479856/83e2b5e0-f035-4403-9611-cc5560fa253f

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01502a04f32100df3d</li>
        <li>Upwork Job ID: 1760726284749012992</li>
        <li>Last Price Increase: 2024-03-07</li>
    </ul>
</details>

<details><summary>Issue Owner</summary>Current Issue Owner: @zanyrenney</details>

<details><summary>Issue Owner</summary>Current Issue Owner: @zanyrenney</details>

**Comment by isagoico** [comment]
We think these 2 might be related but we were not sure. Seems like deeplinking doesn't work twice in a row or just when the app is open in the background.

**Comment by mallenexpensify** [comment]
Thanks @isagoico, adding to #vip-vsb project.

**Comment by zanyrenney** [comment]
Triaged, thanks for adding to #VIPVSB

**Comment by jjcoffee** [comment]
Waiting for proposals!

**Comment by zanyrenney** [comment]
still waiting on proposals.

**Comment by zanyrenney** [comment]
took this here - https://expensify.slack.com/archives/C01GTK53T8Q/p1709567026115789

**Comment by bernhardoj** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Can't sign in using the magic code link as an anonymous user.

### What is the root cause of that problem?
When we open the magic link, it should open the ValidateLogin page. The screen is available on AuthScreens,
https://github.com/Expensify/App/blob/79d6cdafe77879a2aa3bc01d62b5a6a3ac105149/src/libs/Navigation/AppNavigator/AuthScreens.tsx#L277-L285

but the validate login page is not included as the accessible route for an anonymous user,
https://github.com/Expensify/App/blob/79d6cdafe77879a2aa3bc01d62b5a6a3ac105149/src/libs/actions/Session/index.ts#L951-L955

so we get redirected to the sign-in modal page and nothing happens because we are already there.
https://github.com/Expensify/App/blob/79d6cdafe77879a2aa3bc01d62b5a6a3ac105149/src/libs/actions/Report.ts#L2255-L2258
https://github.com/Expensify/App/blob/79d6cdafe77879a2aa3bc01d62b5a6a3ac105149/src/libs/actions/Session/index.ts#L239

The next problem is, that even if we are navigated to the validate login page, the page will immediately close if the user is logged in, even as an anon user. This happens on web too. So, the sign in process is never done.
https://github.com/Expensify/App/blob/79d6cdafe77879a2aa3bc01d62b5a6a3ac105149/src/pages/ValidateLoginPage/index.tsx#L19-L26

After fixing the issue, we will notice there is another issue on this page too that is after successfully validating the magic code from the validate login page, the sign-in modal isn't closed. The code to do the validation already has a navigation logic that is supposed to close the validate login page immediately,
https://github.com/Expensify/App/blob/79d6cdafe77879a2aa3bc01d62b5a6a3ac105149/src/libs/actions/Session/index.ts#L923-L930

but it's not working as expected (as explained [here](https://github.com/Expensify/App/issues/36618#issuecomment-1947913057)). So, we will see a full screen loading indicator until the user is signed-in.

The validate login page will be closed by the sign-in modal dismiss logic itself.
https://github.com/Expensify/App/blob/79d6cdafe77879a2aa3bc01d62b5a6a3ac105149/src/pages/signin/SignInModal.tsx#L27-L30

### What changes do you think we should make in order to solve the problem?

First, include validate login page as an accessible page. 

Validate login route has params, so we can't include `ROUTES.VALIDATE_LOGIN` to `routesCanAccessByAnonymousUser`. Instead, we can use a regex like we did here.
https://github.com/Expensify/App/blob/79d6cdafe77879a2aa3bc01d62b5a6a3ac105149/src/components/DeeplinkWrapper/index.website.tsx#L26-L27

Or we can get the route name and check whether it's "v" (for validate login page) or not.
`const routeName = getStateFromPath(route)?.routes[0].name;`

For the second issue, we should allow anon user to validate login by checking the auth token type.
https://github.com/Expensify/App/blob/79d6cdafe77879a2aa3bc01d62b5a6a3ac105149/src/pages/ValidateLoginPage/index.tsx#L19
`if (session?.authToken && session?.authTokenType !== CONST.AUTH_TOKEN_TYPES.ANONYMOUS) {`

For the third issue, update the navigation to
`Navigation.goBack();`

_if we want to keep showing the full screen loading indicator while signing in, then we should close the validate login page only after the sign in is success instead of doing it immediately_

**Comment by zanyrenney** [comment]
@jjcoffee please can you review to the proposal from @bernhardoj

**Comment by flodnv** [comment]
Sounds good 👍 Let's fix these comments while we're here: https://github.com/Expensify/App/pull/29248#pullrequestreview-1922883996

**Comment by flodnv** [comment]
@zanyrenney ^ 😕

**Comment by bernhardoj** [comment]
> Let's fix these comments while we're here: https://github.com/Expensify/App/pull/29248#pullrequestreview-1922883996

Addressed in the PR.

PR is ready

cc: @jjcoffee

**Comment by jjcoffee** [comment]
@zanyrenney Applied! :pray:

**Comment by bernhardoj** [comment]
Applied

**Comment by zanyrenney** [comment]
offers sent to @jjcoffee and @bernhardoj on UW ahead of payment tomorrow.

**Comment by zanyrenney** [comment]
Hmm sctually the offer to @jjcoffee is timing out : 
![2024-03-18_17-39-21](https://github.com/Expensify/App/assets/56830058/be546872-771c-485a-a4ee-89a28530d87f)

I will do a clean signout tonight and resend.

**Comment by zanyrenney** [comment]
**payment summary**

paid $500 to @bernhardoj 
offer sent to @jjcoffee

**Comment by jjcoffee** [comment]
@zanyrenney Accepted the offer, thanks!

**Comment by zanyrenney** [comment]
payment summary

paid $500 to @bernhardoj
paid $500 to @jjcoffee
